app/models/manageiq/providers/hawkular/middleware_manager/refresh_parser.rb

```diff
@@ -40,7 +42,54 @@ def get_middleware_servers
             @data[:middleware_servers] << server
             @data_index.store_path(:middleware_servers, :by_ems_ref, server[:ems_ref], server)
           end
-        end.flatten
+        end
+      end
+
+      def fetch_domains_with_servers
+        @data[:middleware_domains] = []
+        @data[:middleware_server_groups] = []
+        @ems.feeds.each do |feed|
+          @ems.domains(feed).each do |domain|
+            parsed_domain = parse_middleware_domain(domain)
+            fetch_server_groups(feed, domain)
+
+            # add the server groups to the domain
+            parsed_domain[:middleware_server_groups] = @data[:middleware_server_groups]
+            @data[:middleware_domains] << parsed_domain
+            @data_index.store_path(:middleware_domains, :by_ems_ref, parsed_domain[:ems_ref], parsed_domain)
+
+            # now it's safe to fetch the domain servers
+            fetch_domain_servers(domain)
+          end
+        end
+      end
+
+      def fetch_server_groups(feed, domain)
+        @ems.server_groups(feed, domain).each do |group|
+          parsed_group = parse_middleware_server_group(group)
+          @data[:middleware_server_groups] << parsed_group
+          @data_index.store_path(:middleware_server_groups, :by_name, parsed_group[:name], parsed_group)
+        end
+      end
+
+      def fetch_domain_servers(domain)
+        @ems.child_resources(domain.path).each do |child|
+          next unless child.type_path.end_with?(hawk_escape_id('Domain WildFly Server'))
+          @eaps << child
+
+          server_name = parse_domain_server_name(child.id)
+          server = parse_middleware_server(child, server_name)
+
+          # add the association to server group (it's well hidden in the inventory)
```

Comment:
> it means, that the information about what server is in which server group is not part of the server resource, nor part of the server group resource, nor the domain in the hawkular inventory. But it's well hidden under the server-config resource's configuration - https://github.com/Jiri-Kremser/manageiq/blob/d8d326976f5ba71c7778ff4ae0ba39bcceb23db8/spec/vcr_cassettes/manageiq/providers/hawkular/middleware_manager/refresher.yml#L1551
> 
> It's because the hawkular agent blindly copies the DMR structure from the WildFly into the hawkular inventory, so instead of some useful aggregated data that's easy to use each client has to make sense of this "mess" (ehm, low level data spread across multiple resources that lack any kind of logic :])

if you can distill this great explanation to make the comment in the code more clear, it will be very helpful to whoever will read the code (and won't see these PR comments).